routing/payment_lifecycle.go

```diff
@@ -226,6 +312,64 @@ func (p *paymentLifecycle) resumePayment() ([32]byte, *route.Route, error) {
 type shardHandler struct {
 	paymentHash lntypes.Hash
 	router      *ChannelRouter
+
+	// quit is closed to signal the sub goroutines of the payment lifecycle
+	// to stop.
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// launchAndCollect is an async wrapper around the launch and collect methods,
+// that will send a shard along the goven router and deliver the outcome on the
+// passed channel.
+func (p *shardHandler) launchAndCollect(rt *route.Route,
+	outcomeChan chan *shardOutcome, errChan chan error) {
+
+	// We have a new route to try. Launch a shard along this route.
+	// Since we'll send any outcome async on the channels, we inspect the
+	// outcome and error first in the goroutine below to avoid blocking on
+	// the channel send.
+	attempt, outcome, launchErr := p.launch(rt)
+
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+
+		// Critical launch error.
+		if launchErr != nil {
+			select {
+			case errChan <- launchErr:
+			case <-p.quit:
+			}
+			return
+		}
+
+		// We already have an outcome available, send it and return.
+		if outcome != nil {
+			select {
+			case outcomeChan <- outcome:
+			case <-p.quit:
+			}
+			return
+		}
```

Comment:
The reason is that `launchAndCollect` is called synchronously  from the same loop that reads on that channel, so it will deadlock. The reason I do it sync in the first place is what we discussed earlier, that we want to ensure the shard is fully committed to the link before doing another pathfinding attempt.

I admit it's a bit awkward, open to suggestions for how we can improve it.

Reply:
Ok, clear